controller/node_controller.go

```diff
@@ -995,6 +1003,178 @@ func (nc *NodeController) syncNodeStatus(pod *corev1.Pod, node *longhorn.Node) e
+	for _, version := range brokenKernelVersion {
+		if !strings.Contains(kernelVersion, version) {
+			continue
+		}
+		var kernel, major, minor, patch int
+		if _, err := fmt.Sscanf(kernelVersion, "%d.%d.%d-%d", &kernel, &major, &minor, &patch); err != nil {
+			return errors.Wrap(err, "failed to parse kernel version "+kernelVersion)
+		}
+		if (kernel == majorKernel656Version && patch == brokenKernel656PatchVersion) || (kernel == majorKernel5150Version && (fixedKernel5150PatchVersion > patch && patch >= brokenKernel5150PatchVersion)) {
+			node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeSupportedKernel, longhorn.ConditionStatusFalse,
+				string(longhorn.NodeConditionReasonKernelNotSupport),
+				fmt.Sprintf("Node %v has a kernel version %v known to have a breakage that affects Longhorn. See description and solution at https://longhorn.io/kb/troubleshooting-rwx-volume-fails-to-attached-caused-by-protocol-not-supported", node.Name, kernelVersion))
+			brokenFlag = true
+			break
+		}
+	}
+	if !brokenFlag {
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeSupportedKernel, longhorn.ConditionStatusTrue, "", "")
+	}
+
+	return nil
+}
+
+func (nc *NodeController) syncPackagesInstalled(kubeNode *corev1.Node, node *longhorn.Node, namespaces []lhtypes.Namespace) error {
+	osImage := strings.ToLower(kubeNode.Status.NodeInfo.OSImage)
+	queryPackagesCmd := ""
+	packages := []string{}
+	switch {
+	case strings.Contains(osImage, "ubuntu"):
+	case strings.Contains(osImage, "debian"):
+		queryPackagesCmd = "dpkg -l | grep -w"
+		packages = []string{"nfs-common", "open-iscsi", "cryptsetup", "dmsetup"}
+	case strings.Contains(osImage, "centos"):
+	case strings.Contains(osImage, "fedora"):
+	case strings.Contains(osImage, "rocky"):
+	case strings.Contains(osImage, "ol"):
+		queryPackagesCmd = "rpm -q"
+		packages = []string{"nfs-utils", "iscsi-initiator-utils", "cryptsetup", "device-mapper"}
+	case strings.Contains(osImage, "suse"):
+		queryPackagesCmd = "rpm -q"
+		packages = []string{"nfs-client", "open-iscsi", "cryptsetup", "device-mapper"}
+	case strings.Contains(osImage, "arch"):
+		queryPackagesCmd = "pacman -Q"
+		packages = []string{"nfs-utils", "open-iscsi", "cryptsetup", "device-mapper"}
+	case strings.Contains(osImage, "gentoo"):
+		queryPackagesCmd = "qlist -I"
+		packages = []string{"net-fs/nfs-utils", "sys-block/open-iscsi", "sys-fs/cryptsetup", "sys-fs/lvm2"}
+	default:
+		node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeRequiredPackages, longhorn.ConditionStatusFalse,
+			string(longhorn.NodeConditionReasonUnknownOS),
+			fmt.Sprintf("Node %v has an unknown OS image %v, please make sure the necessary packages are installed", node.Name, osImage))
+		return nil
+	}
+
+	nsexec, err := lhns.NewNamespaceExecutor(lhtypes.ProcessNone, lhtypes.HostProcDirectory, namespaces)
+	if err != nil {
+		return err
```

Comment:
We can log the error to NodeConditionTypeRequiredPackages condition.

Reply:
Logged the error to NodeConditionTypeRequiredPackages condition. Thanks